lib/diff.ml

```diff
@@ -101,13 +114,37 @@ and module_item ~typing_env ~name ~reference ~current =
       module_declaration ~typing_env ~name ~reference ~current
   | _ -> assert false
 
+and module_type_item ~typing_env ~name ~reference ~current =
+  match (reference, current) with
+  | None, None -> None
+  | None, Some (ModType curr_mtd) ->
+      Some (Modtype { mtname = name; mtdiff = Added curr_mtd })
+  | Some (ModType ref_mtd), None ->
+      Some (Modtype { mtname = name; mtdiff = Removed ref_mtd })
+  | Some (ModType ref_mtd), Some (ModType curr_mtd) ->
+      module_type_declaration ~typing_env ~name ~reference:ref_mtd
+        ~current:curr_mtd
+  | _ -> assert false
+
 and module_declaration ~typing_env ~name ~reference ~current =
-  match (reference.md_type, current.md_type) with
+  module_helper ~typing_env ~name ~ref_module_type:reference.md_type
+    ~current_module_type:current.md_type ~reference_location:reference.md_loc
+
+and module_type_declaration ~typing_env ~name ~reference ~current =
+  match (reference.mtd_type, current.mtd_type) with
+  | Some ref_sub, Some curr_sub ->
+      module_helper ~typing_env ~name ~ref_module_type:ref_sub
+        ~current_module_type:curr_sub ~reference_location:reference.mtd_loc
+  | _ -> assert false
+
+and module_helper ~typing_env ~name ~ref_module_type ~current_module_type
```

Comment:
I think we can come up with a better name here. If we rename `modtype` to `module_type_fallback` and `module_helper` to `module_type` then it's both clearer what those functions are for and closer to the naming convention of the other functions in this module.

What do you think?